### PR TITLE
Initial work to support advanced compilation of UMD bundles using GCC

### PR DESCRIPTION
**Pull request by banga**
As described in #11092, this PR adds support for passing in externs to GCC to allow it to compile bundles in advanced mode without renaming properties that are part of the public API. Most of the externs were taken from https://github.com/cljsjs/packages/tree/master/react/resources/cljsjs/, although some more needed to be added. I've split them into various files with necessary comments so they should be reasonably easy to follow.

At this point, the node bundles do not build correctly (the module.exports part is stripped out), so we can't run tests just yet. So I just tested by pointing the babel-standalone fixture to the built prod files and loading it. The current fixture works successfully, and I was able to create a functional component as well. But when I tried creating a component by extending `React.Component`, I saw `class constructor cannot be invoked without 'new'`, indicating that another property is getting renamed and causing the functional component codepath to execute for non-functional components.

(This is my first PR to an FB project, so I just submitted the CLA.)

- [ ] Get all `test-build-prod` tests to pass
- [ ] Don't use fbjs externs for UMD bundles for better minification
- [ ] Clean up externs files to remove old methods that no longer exist
- [ ] Try [Splitting JS into modules](https://stackoverflow.com/questions/10395810/how-do-i-split-my-javascript-into-modules-using-googles-closure-compiler/10401030#10401030) to allow minifying react internal properties accessed across bundle boundaries

### CODE REVIEW

**Comment by sophiebits** [comment]
That error is likely caused by renaming isReactComponent.

File: externs/react.ext.js [hunk 1]
```diff
@@ -0,0 +1,1756 @@
...
+var React = {};
+
+/**
+ * @type {string}
+ * @const
+ */
+React.version;
+
+React.createClass = function(specification) {};
+React.createFactory = function(reactClass) {};
```
Comment (gaearon) [inline externs/react.ext.js]: A lot of these are outdated. Once we get the tests passing we should trim these down significantly.

**Comment by banga** [comment]
@sophiebits yeah that was it. I got the node bundles to build correctly last night and got a handful of `test-build-prod` tests to pass. I'll clean-up my changes and push tonight. The most frequent error that was breaking tests now was `Target Container is not a DOM Element`, which might be too generic to debug just from that, but let me know if you have ideas.

**Comment by gaearon** [comment]
Can't wait 😍

**Comment by appsforartists** [comment]
The [Angular team](https://github.com/angular) has tools ([tsickle](https://github.com/angular/tsickle) and [clutz](https://github.com/angular/clutz)) to convert between TypeScript and Closure annotations.  I haven't tested them myself yet, so I don't know how well they work, but I believe the goal is to enable TypeScript applications to use advanced GCC.

It may be worth exploring using [tsickle](https://github.com/angular/tsickle) to auto-gen your Closure externs.  I'm not sure if it would work on Flow code, but I bet @evmar is amenable to PRs.

I suspect that automatically generating them is better in the long run than maintaining the externs by hand.

**Comment by gaearon** [comment]
One thing that’s nice in our case is that the API surface is relatively small. So once we get them right for the first time they’re not as much pain to maintain.

**Comment by banga** [comment]
Alright after today's work, we have 222 out of 244 tests passing in the `react` package:
image.

**Comment by evmar** [comment]
@appsforartists Within Google we do generate externs for React from the [DefinitelyTyped React d.ts files](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts).  The API surface is a bit larger than you'd expect due to including all the CSS properties etc.

But it's also probably more complicated than what you want because:
1) the d.ts files include lots of type definitions (including fancy generics to model `setState()` etc)
2) they ended up needing some fixing by hand at the end due to various bugs (in how we handle said fancy generics -- Closure can't model some of them)
3) if you're just worrying about the untyped Closure optimizations you don't need the types.

**Comment by banga** [comment]
Yeah, the more challenging part seems to be the non-public API shared between the bundles (e.g. `FiberNode` and `updater`). We need to either prevent them from being renamed, or find a way to compile everything together and split the bundles so they are renamed consistently. So far I'm just manually doing that, with some hints from exported types.

**Comment by stereobooster** [comment]
When Dan twitted about using GCC for minification I thought about "GNU Compiler Collection", but this was about "Google Closure Compiler" :rofl:

**Comment by banga** [comment]
Ok now all `react` and `react-dom` tests are passing! Only 18 more to go...

**Comment by gaearon** [comment]
We probably don’t need to apply advanced mode to any bundles except React and ReactDOM. Their size is less critical / important.

**Comment by gaearon** [comment]
It’s also worth exploring a subset of this PR that only applies advanced to the React bundle. It’s not as important as ReactDOM but it’s a good way to get started since it’s fairly small and won’t need a ton of externs. That would be relatively safe to merge while we continue to iterate on ReactDOM.

**Comment by sophiebits** [comment]
If we do run it on all bundles, let’s have separate externs for each bundle. There are some names we want to munge in react-dom but not react-reconciler, for instance.

**Comment by banga** [comment]
@gaearon sounds good. Should I start a separate PR for just compiling the react bundle?

@sophiebits yup, I’ve been keeping the externs separate by bundle type, although some bundles need to pull in externs for others (eg react-dom needs react externs)

**Comment by gaearon** [comment]
Yes, a separate more scoped PR that can pass a CI would be great. Thank you for working on this!

**Pull request closed without merging**